src/Microsoft.Extensions.Configuration.AzureAppConfiguration/FeatureManagement/FeatureManagementKeyValueAdapter.cs

```diff
@@ -195,18 +197,60 @@ public FeatureManagementKeyValueAdapter(FeatureFilterTracing featureFilterTracin
 
                 string telemetryPath = $"{featureFlagPath}:{FeatureManagementConstants.Telemetry}";
 
-                if (telemetry.Enabled)
-                {
-                    keyValues.Add(new KeyValuePair<string, string>($"{telemetryPath}:{FeatureManagementConstants.Enabled}", telemetry.Enabled.ToString()));
-                }
-
                 if (telemetry.Metadata != null)
                 {
                     foreach (KeyValuePair<string, string> kvp in telemetry.Metadata)
                     {
                         keyValues.Add(new KeyValuePair<string, string>($"{telemetryPath}:{FeatureManagementConstants.Metadata}:{kvp.Key}", kvp.Value));
                     }
                 }
+
+                if (telemetry.Enabled)
+                {
+                    byte[] featureFlagIdHash;
+
+                    using (HashAlgorithm hashAlgorithm = SHA256.Create())
+                    {
+                        featureFlagIdHash = hashAlgorithm.ComputeHash(Encoding.UTF8.GetBytes($"{setting.Key}\n{(string.IsNullOrWhiteSpace(setting.Label) ? null : setting.Label)}"));
+                    }
+
+                    string featureFlagIdBase64 = Convert.ToBase64String(featureFlagIdHash);
+
+                    int indexOfEquals = featureFlagIdBase64.IndexOf("=");
+
+                    int stringBuilderCapacity = indexOfEquals != -1 ? indexOfEquals : featureFlagIdBase64.Length;
+
+                    StringBuilder featureFlagIdBuilder = new StringBuilder(stringBuilderCapacity);
+
+                    for (int i = 0; i < stringBuilderCapacity; i++)
+                    {
+                        if (featureFlagIdBase64[i] == '+')
+                        {
+                            featureFlagIdBuilder.Append('-');
+                        }
+                        else if (featureFlagIdBase64[i] == '/')
+                        {
+                            featureFlagIdBuilder.Append('_');
+                        }
+                        else
+                        {
+                            featureFlagIdBuilder.Append(featureFlagIdBase64[i]);
+                        }
+                    }
```

Comment:
Can we add some comments to explain how the featureFlagId is supposed to be calculated? It's not clear why we are doing some of the operations we are doing here.

Reply:
I think it would also add clarity if we move the b64url into an extension method.

ToBase64UrlString(this bytes[]);

At that point the code will be more self explanatory.

Reply:
Right, won't mind if we move the whole featureFlagId calculation to a separate method. This function is getting way too long.

Reply:
I think I'll go with the extension for now since it makes sense to me conceptually, but let me know if you have any other concerns

Reply:
I just want to point out that the featureFlagId calculation is "by convention". It's not persisted anywhere. Portal and all our other providers will do the same calculation on-fly. Since the .NET provider is the first one doing this, it will make everyone's life easier if we can encapsulate it in one function with detailed comments so others can review/reference.

Reply:
@amerjusupovic Can we please go with b64url extension method and private static method for `CalculateFeatureFlagId(string key, string label)`

Reply:
sounds good to me

Reply:
@zhenlan do you think the extension method is unnecessary? I thought it made sense to be an extension just because it's a known conversion, but it's true that we're only using it here.

Reply:
I would recommend keeping that extension method. b64url is a spec and encapsulates well as utility method.